### PR TITLE
Flush instead of synced-flush inactive shards

### LINKED ISSUE #31965: Flush inactive shards
We currently have a logic that triggers a sync flush when a primary shard becomes inactive (after 5 minutes of no write activity on the primary shard). The goal of this is to ensure that sync flush markers are in place after a period of inactivity, so that a full cluster / rolling restart of nodes results in quick peer recoveries when there is no write activity on the respective shard. With operation-based recoveries, we also provide fast recoveries when there is write activity during node restarts. Operation-based recovery can, however, more frequently trigger situations where a replica shard becomes inactive, yet not all its searchable segments are flushed to disk, as the flushing is only triggered when a primary becomes inactive, and is not triggered by subsequent recoveries of replicas. This results in unnecessary extra storage (more translog generations + more Lucene segments)  and possibly slows down future store- and peer-based recoveries. /cc: @jpountz 

The following test illustrates the issue:

```
package org.elasticsearch.indices.flush;

import org.elasticsearch.action.admin.indices.segments.IndexShardSegments;
import org.elasticsearch.action.admin.indices.segments.ShardSegments;
import org.elasticsearch.client.Client;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.indices.IndexingMemoryController;
import org.elasticsearch.test.ESIntegTestCase;
import org.elasticsearch.test.InternalTestCluster;

import java.util.List;

import static org.hamcrest.Matchers.equalTo;

@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
public class FlushOnInactivityIT extends ESIntegTestCase {

    public void testFlushOnInactivity() throws Exception {
        List<String> nodes = internalCluster().startNodes(2,
            Settings.builder().put(IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING.getKey(), "3s").build());

        client().admin().indices().prepareCreate("test").get();

        ensureGreen("test");

        index("test", "_doc", "1");
        refresh("test"); // create segment
        index("test", "_doc", "2");
        refresh("test"); // create segment

        internalCluster().restartNode(nodes.get(0), new InternalTestCluster.RestartCallback() {

            public Settings onNodeStopped(String nodeName) throws Exception {
                assertBusySegmentsFlushed(client(nodes.get(1)), "test");
                return super.onNodeStopped(nodeName);
            }

        });

        ensureGreen("test");

        assertBusySegmentsFlushed(client(), "test");
    }

    private void assertBusySegmentsFlushed(Client client, String index) throws Exception {
        assertBusy(() -> {
            for (IndexShardSegments indexShardSegments : client.admin().indices().prepareSegments(index).get().getIndices().get(index)
                .getShards().values()) {
                for (ShardSegments shardSegments : indexShardSegments) {
                    assertThat(shardSegments.getNumberOfCommitted(), equalTo(shardSegments.getNumberOfSearch()));
                }
            }
        });
    }

}
```

### PR DESCRIPTION
**Pull request by dnhatn**
With peer recovery retention leases and sequence-number based replica allocation, a regular flush can speed up recovery as a synced-flush. With this change, we will flush instead of synced-flush when a shard becomes inactive.

Depends on #48430
Closes #31965

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
I think we will still need the old sync-flush logic in 7.x to handle 6.x/7.x indices without soft-deletes as rolling/full cluster restarts are o.w. going to be painful.

**Comment by dnhatn** [comment]
> I think we will still need the old sync-flush logic in 7.x to handle 6.x/7.x indices without soft-deletes as rolling/full cluster restarts are o.w. going to be painful.

Yes, we should keep this logic in 7.x.

**Review by ywelsch — APPROVED** [review]
LGTM. I think the flushing behavior should be backported to 7.x (and the sync-flush behavior kept in 7.x as well for non-soft-deletes indices)

File: server/src/test/java/org/elasticsearch/indices/flush/FlushIT.java [hunk 1]
```diff
@@ -369,4 +382,29 @@ public void testDoNotRenewSyncedFlushWhenAllSealed() throws Exception {
         assertThat(forthSeal.successfulShards(), equalTo(numberOfReplicas + 1));
         assertThat(forthSeal.syncId(), not(equalTo(thirdSeal.syncId())));
     }
+
+    public void testFlushOnInactive() throws Exception {
+        final String indexName = "flush_on_inactive";
+        List<String> dataNodes = internalCluster().startDataOnlyNodes(2, Settings.builder()
+            .put(IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING.getKey(), randomTimeValue(2, 5, "s")).build());
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/indices/flush/FlushIT.java]: Could we maybe make the setting smaller to ensure the test completes faster? Maybe use:

```
TimeValue.timeValueMillis(scaledRandomIntBetween(10, 2000))
```
instead?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -1758,19 +1758,28 @@ public void addShardFailureCallback(Consumer<ShardFailure> onShardFailure) {
 
     /**
      * Called by {@link IndexingMemoryController} to check whether more than {@code inactiveTimeNS} has passed since the last
-     * indexing operation, and notify listeners that we are now inactive so e.g. sync'd flush can happen.
+     * indexing operation, so we can flush the index.
      */
     public void checkIdle(long inactiveTimeNS) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: ++. Addressed in https://github.com/elastic/elasticsearch/pull/49126/commits/ecba8ae620f40e435d9ec60575b1a16e458400c0.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -1758,19 +1758,28 @@ public void addShardFailureCallback(Consumer<ShardFailure> onShardFailure) {
...
+                    @Override
+                    public void onFailure(Exception e) {
+                        if (state != IndexShardState.CLOSED) {
+                            logger.warn("failed to flush shard on inactive", e);
+                        }
+                    }
+
+                    @Override
+                    protected void doRun() {
+                        periodicFlushMetric.inc();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Yes, I pushed https://github.com/elastic/elasticsearch/pull/49126/commits/a873924023eaf32e825a3566de46f18d895bff0f to increase the metric after flushing.

File: server/src/test/java/org/elasticsearch/indices/flush/FlushIT.java (hunk 1 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/indices/flush/FlushIT.java]: Adjusted in https://github.com/elastic/elasticsearch/pull/49126/commits/330c108b1038d2f18516fb4014d2068d29bf3d09.

**Comment by dnhatn** [comment]
Thanks @ywelsch and @henningandersen for reviewing.

**Pull request merged**